### DISCUSSION TITLE
App router can't answer anything but http status 200

### DISCUSSION THREAD

**Discussion by iliaschaumont** [vercel/next.js#70170] [App Router]
### TL;DR : can't manage to return anything but 200. Got workaround below but definitely prefer cleaner option

Hello

I use the app router with Next14. I have a dynamic route based on the id of items in my DB. I do Server Side fetching, so I do my data fetching in the export default of my file /app/[locale]/show/[id]/page.jsx
After I do getDataFromDB(id) I would like to return a 404 response if I didn't get anything because users and other softwares such as Google SEO should know "not everything went well". I use the "notFound()" function from next/navigation.
When I look at my app logs I always see that my frontend answers with a 200, even if the url doesn't make any sense ("localhost:8081/en/show/thisisreally444notan234id/ANDEOHVEOZFO").

User see the 404 page from my not-found.jsx file, but the http code is always 200. I used postman to check, and the http code is truly always 200.

<img width="749" alt="Capture d’écran 2024-09-17 à 01 21 39" src="https://github.com/user-attachments/assets/6053882d-b0d3-43d3-8ba6-2ef3b0c02733">

I really struggle with this because that means I can't provide a relevant status code and my SEO is dead because I can't return proper 404 or 410.
Any way to do this with app router ? What hidden gem in the documentation or discussions did I miss ?

### Considered workarounds:
1. Switch to app router --> really not worth it, too much work
2. In the middleware --> I'd have to parse every incoming request before even routing, see if it matches the route, isolate the ID and do a DB query inside the middleware. It might work, yes, but I'm not very comfy with calling my services in the middleware. Plus, it would make queries roughly twice as long .
3. Forget about 400 status code. --> I leverage my own error management in server rendering and if I don't get data, I render a special component instead of the page ("sorry, something happened or doesn't exist..."). Google robots would still try to index these pages and would get 200 as a response, but I'll have to be patient and wait for them to be buried deep. No one will find them anyway as the metadata will be empty or placeholders. I don't like hiding errors, but this might be the smoothest ?
4. Alternative to 3. --> when no data found on the server side, use permanent redirect to the homepage. Not extremely clean, but at least Google wouldn't index thousands of useless pages.

Thanks for giving any help or tips 🙏 that'd be very appreciated !

**Comment by iliaschaumont** [comment]
I reproduce with a very small configuration : 

```
import { notFound } from 'next/navigation'

export default function CatchAllPage() {
	if (true) {
		notFound()
	}

	return <div>Hello this is my page</div>
}

```
A catchall route that returns notFound all the time. It displays the 404 page but still returns 200 code.
This thing is driving me crazy

**Comment by iliaschaumont** [comment]
Alternatively, tried others next/navigation methods to see if they work and... I'm afraid not as documentation says.
```
import { notFound, permanentRedirect } from 'next/navigation'

export default function CatchAllPage() {
	if (true) {
		permanentRedirect('/en/about')
	}

	return <div>Hello this is my page</div>
}
```

When accessing localhost:8081/en/foo I am redirected to /en/about BUT with a 200 status code.
Same thing with curl or Postman.

Documentation says these methods should handle specific status codes so I'm very confused
